Dashboard (ILIAS 9): Feature: Global Dashboard Settings of Presentation and Sortation for all Dashboard Sections

fhelfer opened the pull request:
Feature Wiki: https://docu.ilias.de/goto_docu_wiki_wpage_7504_1357.html

fhelfer committed:
move favourites and courses&groups tab

fhelfer committed:
refactor dashboard sortation and presentations

fhelfer committed:
license dashboard

mjansenDatabay commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -24,10 +27,11 @@
 {
-    protected \ILIAS\UI\Factory $ui_factory;
-    protected \ILIAS\UI\Renderer $ui_renderer;
+    private ilRbacSystem $rbacsystem;
```
`rbac_system` is available in the base class.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -156,7 +169,23 @@ public function initForm() : \ILIAS\UI\Component\Input\Container\Form\Standard
+        return $this->ui_factory->input()->container()->form()->standard(
+            $this->ctrl->getFormAction($this, 'savePresentation'),
+            array_map(
+                function (int $view) {
```
Please add a return type.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +241,93 @@ public function setSettingsSubTabs(string $a_active) : void
-        // presentation
         $ops = $this->viewSettings->getAvailablePresentationsByView($view);
-        $pres_options = array_column(array_map(function ($k, $v) use ($lng) {
+        $pres_options = array_column(array_map(static function ($k, $v) use ($lng) {
```
Please add parameter and return types.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +241,93 @@ public function setSettingsSubTabs(string $a_active) : void
+    {
+        $lng = $this->lng;
+        $ops = $this->viewSettings->getAvailableSortOptions();
+        $sortation_options = array_column(array_map(static function ($k, $v) use ($lng) {
```
Please add parameter and return types.

lscharmer commented on Services/Dashboard/ItemsBlock/classes/class.ilPDSelectedItemsBlockViewSettings.php:
```diff
@@ -1,70 +1,59 @@
-    /**
-     * @var array[]
-     */
+    /** @var array[] */
```
Please make the type annotation of `$availablePresentationsByView` more specific: `array[]` -> `array<int, string[]>`

fhelfer committed:
add new classes for learning sequence dashboard block

fhelfer committed:
dashboard sorted language files

fhelfer committed:
copyright dashboard

fhelfer committed:
dashboard administration sortation sub tab settings

fhelfer committed:
new abstract class DashboardBlockGUi for PD Panels

fhelfer committed:
new panels dashboard extending new dasboardblockui

fhelfer committed:
implementing dashboard panels

fhelfer committed:
new dashboardblock gui manage object

iszmais commented:
@fhelfer Just a short state check

Are you getting forward?

Pls respond if you already should do some prereview.

iszmais commented:
@fhelfer Nur ein kurzes State check. 

Kommst du gut voran?

Sag gerne bescheid wenn wir schonmal was prereviewn sollen.

iszmais commented:
@fhelfer Nur ein kurzes State check. 

Kommst du gut voran?

Sag gerne bescheid wenn wir schonmal was prereviewn sollen.

fhelfer commented:
Hey @iszmais,

ja klappt ganz gut, einiges ist schon implementiert und auf die neue ilBlock Klasse umgestellt.
Ihr könnt gern schonmal reingucken ob sich das so deckt mit dem was ihr euch vorgestellt habt
Ihr habt sicher Verbesserungen ;-)

iszmais commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -176,12 +176,12 @@ public function setOffset(int $a_offset): void
     public function getOffset(): int
     {
-        return $this->offset;
+        return $this->offset ?? 0;
```
Better use a default in the definition (top).
Getter should always return the exact value (or an error).

iszmais commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -176,12 +176,12 @@ public function setOffset(int $a_offset): void
     public function correctOffset(): void
     {
-        if (!($this->offset < $this->max_count)) {
+        if (!(($this->offset ?? 0) < $this->max_count)) {
```
Im not sure if this is used somewhere else but i would recommend to put that functionality inside the setter.
Youmay return an Error also if the checks dont match. Just setting it 0 would be kinda an silencer. Or is there a case where an invalid input is expected.

Just like that for Example:
```
    public function setOffset(int $a_offset): void
    {
        if($this->checkOffset($a_offset)){
            $this->offset = $a_offset;
        } else {
             thrown new ...
        }
    }

    private function checkOffset(int $a_offset): bool
    {
         return $this->offset < $this->max_count;
    }
```

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -67,7 +71,7 @@ public function executeCommand(): void
         $this->prepareOutput();
 
-        if (!$this->rbacsystem->checkAccess("visible,read", $this->object->getRefId())) {
+        if (!$this->rbac_system->checkAccess("visible,read", $this->object->getRefId())) {
```
ILIAS usallly uses rbacsystem as naming.
But im not quite sure where this comes from here in the first place...

fhelfer replied:
declared in ilObjectGUI, changing it would require it to be adapted in all inheriting classes

iszmais replied:
Ok than just leave it to that. We just resolving our own issues.

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,60 @@ public function editSettings(): void
+        $options = array_reduce(
+            $availabe_sort_options,
+            function (array $options, string $option): array {
+                $options[$option] = $this->lng->txt("dash_sort_by_" . $option);
```
In most parts its better you put static string instant a constant on top of the class, especially when you use the multiple time. This may not always be the best solution but in most cases-

For very generic string from a bigger scope (like 'view') you may search if there is already a constant in ILIAS defined. Just be sure its from the Right context and not only "fitting".
(e.G. there are sometimes double Constant where one is used for the Dashboard and one for the Menu or for the Repository)

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,526 @@
+            $sortingCommands[] = [
+                'txt' => $this->lng->txt('dash_sort_by_' . $sorting),
+                'url' => $this->ctrl->getLinkTarget($this, 'changePDItemSorting'),
+                'xxxasyncUrl' => $this->ctrl->getLinkTarget($this, 'changePDItemSorting', '', true),
```
xxx?

iszmais commented on Services/Dashboard/Block/classes/class.ilRecommendedContentBlockGUI.php:
```diff
@@ -0,0 +1,116 @@
+        $item_gui = $list_factory->byType($data['type']);
+        ilObjectActivation::addListGUIActivationProperty($item_gui, $item);
+
+        $this->ctrl->setParameterByClass(self::class, "item_ref_id", $data['ref_id']);
```
Im not quite sure if you want to use **self::class** here

In general the are 3 types to get a class: `self::class` , `__CLASS__` und `getclass($this)`

Only `getclass($this)` uses inheritance, the other 2 will always refer to your class, even in upcoming child classes.

Therefore you should ask yourself what you want other developers who might wanna use your class and inherit from it to get out there.

Some more info here: https://www.php.de/forum/webentwicklung/php-einsteiger/89234-self-vs-__class__

iszmais commented on lang/ilias_en.lang:
```diff
@@ -7938,6 +7938,11 @@ dash#:#dash_make_favourite#:#Add to Favourites
 dash#:#dash_member_main_alt#:#Courses and groups can also be configured as a separate main menu entry.
 dash#:#dash_memberships#:#My Courses and Groups
 dash#:#dash_presentation#:#Presentation
+dash#:#dash_presentation_favorites#:#Favorites
```
Some English translations are missing (or german translations are obsolete).

And rember to sort the lang file with the fixer (if you havent already).

iszmais reviewed (commented):
LGTM. Just some minor adjustments

fhelfer committed:
bugfix dashboard actor selectable presentation

fhelfer committed:
sortation by location,type,alphabet on dashboard

fhelfer committed:
bugfix activatable sorataions dashboard administration

fhelfer committed:
cleanup old block guis dashboard

bugfix favoriteslistgui + copyright fix

fhelfer committed:
bugfix favorites mainbar + manage Favourites

fhelfer committed:
favourites emptyHandling

fhelfer committed:
mymemberships emptyhandling dashboard

fhelfer committed:
js code to make sure there is always a default dashboard block sortation

fhelfer committed:
bugfix old default views

fhelfer committed:
bugfix initial sorations selected

iszmais reviewed (approved):
LGTM. Just on small comment (s.a.) but thats not merge relevant.

iszmais reviewed (approved):
LGTM. Just on small comment (s.a.) but thats not merge relevant.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+                $this->viewSettings->getActiveSortingsByView($view)
+            )
+            ->withAdditionalOnLoadCode(function ($id) use ($view) {
+                return "$('#$id').attr('data-checkbox', 'activeSorting$view');
```
Please extract this code into an own file.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+                return "$('#$id').attr('data-checkbox', 'activeSorting$view');
+                $('#$id').change(function () {
+                    // not checked options
+                    var disabledOptions = $(this).find('input[type=\"checkbox\"]:not(:checked)');
```
Please use `const` instead of `var`.
Please replace all jQuery calls to plain JS calls, because all of them can be replaced with one function (There are cases where the jQuery calls correspond to multiple vanilla JS calls but this is not the case in for this code).

`$(this).find(cssSelector)` => `this.querySelectorAll(cssSelector)`

`$(this).filter(cssSelector)` => `checkedOptions` and `selectionOptions` are only used once so it can be replaced with `(...).querySelectorAll`:
`disabledOptions.filter('[value=\"' + $(this).val() + '\"]').length > 0` to `this.querySelectorAll('input[type=\"checkbox\"][value=\"' + this.value + '\":not(:checked)]').length > 0`

`$(...).val()` => `(...).value`

`$(...).each` => `(...).forEach` or `Array.from(...).forEach`

`$(...).attr(key, val)` => `(...).setAttribute(key, val)`

`$(...).attr(key)` => `(...).getAttribute(key)`

`$(...).removeAttr` => `(...).removeAttribute(key, value)`

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+                    var selectionInput = $('[data-select=\"sorting$view\"]');
+                    selectionInput.find('option').each(function () {
+                        if (disabledOptions.filter('[value=\"' + $(this).val() + '\"]').length > 0) {
+                        console.log('disbale option' + $(this).val());
```
Please use `console.debug` for information like this.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+                    // if only one option is not disabled make it readonly
+                    if ($(this).find('input[type=\"checkbox\"]:checked').length === 1) {
+                    console.log('make readonly option');
+                       $(this).find('input[type=\"checkbox\"]:checked').attr('disabled', 'disabled');
```
Please use a variable instead of re-selecting the same object as in line 196.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+            ->withRequired(true)
+            ->withAdditionalOnLoadCode(function ($id) use ($view) {
+                // write id to attr
+                return "$('#$id').attr('data-select', 'sorting$view');
```
Please consider moving this code into the same JS file as the on change code above.
Same as the previous JS block applies here.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+            ->select($this->lng->txt("dash_default_sortation"), $options)
+            ->withValue($this->viewSettings->getDefaultSortingByView($view))
+            ->withRequired(true)
+            ->withAdditionalOnLoadCode(function ($id) use ($view) {
```
Please add `static function` and consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+                // write id to attr
+                return "$('#$id').attr('data-select', 'sorting$view');
+                // disable options that are not checked
+                var disabledOptions = $('[data-checkbox=\"activeSorting$view\"]').parent().find('input[type=\"checkbox\"]:not(:checked)');
```
`$(...).parent()` => `(...).parentNode`

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+            ->withValue(
+                $this->viewSettings->getActiveSortingsByView($view)
+            )
+            ->withAdditionalOnLoadCode(function ($id) use ($view) {
```
Please add `static function` and consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+        return $this->ui_factory->input()->container()->form()->standard(
+            $this->ctrl->getFormAction($this, 'saveSorting'),
+            array_map(
+                function (int $view): ILIAS\UI\Component\Input\Field\Section {
```
Please consider using `fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,112 @@ public function editSettings(): void
+        $availabe_sort_options = $this->viewSettings->getAvailableSortOptionsByView($view);
+        $options = array_reduce(
+            $availabe_sort_options,
+            function (array $options, string $option): array {
```
Please add `static function`.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -160,14 +279,30 @@ public function initForm(): \ILIAS\UI\Component\Input\Container\Form\Standard
+        return $this->ui_factory->input()->container()->form()->standard(
+            $this->ctrl->getFormAction($this, 'savePresentation'),
+            array_map(
+                function (int $view): ILIAS\UI\Component\Input\Field\Section {
```
Please consider using `fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +341,100 @@ public function setSettingsSubTabs(string $a_active): void
-        // presentation
         $ops = $this->viewSettings->getAvailablePresentationsByView($view);
-        $pres_options = array_column(array_map(function ($k, $v) use ($lng) {
+        $pres_options = array_column(array_map(static function (int $k, string $v) use ($lng): array {
```
Please consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+    public function setData(array $a_data): void
+    {
+        // unset empty keys
+        $this->data = array_filter($a_data, static function ($a) {
```
Please use `array_filter($a_data)` without a callback it uses `empty`.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+            return strcmp($left['title'], $right['title']);
+        });
+
+        uasort($groups['not_dated'], static function ($left, $right) {
```
Please consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+            }
+        }
+
+        uasort($groups['upcoming'], static function ($left, $right) {
```
Please delete the duplicated callback code:
`ongoing` and `ended` are exactly the same and `upcoming` is the same except that the date comparison is reversed. Please create a callback which can be used for all three `uasort`'s. With a parameter like `$reverse_date_sorting` it can be made directly clear that the order is reversed in one case (switching between `1` and `-1` does not make this clear.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+        }
+
+        foreach ($grouped_items as $key => $group) {
+            usort($grouped_items[$key], static function ($left, $right) {
```
Same callback as in line 160. Please extract this callback into a method like `sortByTitle`.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+        $data = $this->getData();
+        $data = array_merge(...array_values($data));
+
+        $parent_ref_ids = array_values(array_unique(array_map(function ($item) {
```
Please consider using `fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+        }
+        ksort($grouped_items);
+        foreach ($grouped_items as $key => $group) {
+            usort($grouped_items[$key], static function ($left, $right) {
```
Same callback as in line 160. Please extract this callback into a method like sortByTitle.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+                break;
+
+            default:
+                if (method_exists($this, $cmd)) {
```
Please remove this `if` I don't think we need 2 branches for the same purpose and IMO it's already not nice to be able to call private methods with this but with the `Object` postfix it at least provides some basic levels of encapsulation.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+            case ilPDSelectedItemsBlockConstants::SORT_BY_ALPHABET:
+                $data = $this->getData();
+                $data = array_merge(...array_values($data));
+                uasort($data, static function ($a, $b) {
```
Same callback as in line 160. Please extract this callback into a method like sortByTitle.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+        }
+
+        if (is_array($groupedCommands[0])) {
+            $actions = array_map(static function ($item) use ($ui) {
```
Please consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,744 @@
+        }
+
+        $full_class = 'ilObj' . $class . 'ListGUI';
+        require_once $location . '/class.' . $full_class . '.php';
```
I don't think this is necessary anymore because the composer should take care of autoloading.

iszmais reviewed (changes requested):
Noch etwas. Vor kurzem hat sich im Codestyle etwas geändert, sodass LICENSE Header nur vor die declares geschrieben werden. Also nicht mehr
```
<?php

declare(strict_types=1);

/**
 * This file contains an example of coding styles.
 */

namespace Vendor\Package;
```
sondern nun
```
<?php

/**
 * This file contains an example of coding styles.
 */

declare(strict_types=1);

namespace Vendor\Package;
```

Siehe https://discord.com/channels/819582183290437642/938352958876286988/1070290923440848896
und https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements

@fhelfer Bitte noch in allen betroffenen Files ändern.

## FYI

@lscharmer @mbecker-databay @gvollbach @mjansenDatabay @nmatuschek

lscharmer commented on Services/Dashboard/ItemsBlock/classes/class.ilPDSelectedItemsBlockViewSettings.php:
```diff
@@ -162,87 +195,65 @@ public function getSortByAlphabetMode(): string
+        } else {
+            $active_sortings = unserialize($val, ['allowed_classes' => false]);
+        }
+        return array_filter($active_sortings, function ($sorting) use ($view) {
```
Please consider using `fn` (if it can be formatted nicely).

lscharmer commented on Services/Repository/GlobalScreen/classes/RepositoryMainBarProvider.php:
```diff
@@ -118,33 +118,31 @@ public function getStaticSubItems(): array
+                                    ->withSupportsAsynchronousLoading(true)
+                                    ->withTitle($title)
+                                    ->withSymbol($icon)
+                                    ->withContentWrapper(function () use ($f): Legacy {
```
Please add `static function` and consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Repository/GlobalScreen/classes/RepositoryMainBarProvider.php:
```diff
@@ -118,33 +118,31 @@ public function getStaticSubItems(): array
+                                    ->withParent(StandardTopItemsProvider::getInstance()->getPersonalWorkspaceIdentification())
+                                    ->withPosition(10)
+                                    ->withAvailableCallable(
+                                        function () use ($dic): bool {
```
Please add `static function` and consider using `static fn` (if it can be formatted nicely).

lscharmer commented on Services/Repository/GlobalScreen/classes/RepositoryMainBarProvider.php:
```diff
@@ -118,33 +118,31 @@ public function getStaticSubItems(): array
+                                    ->withVisibilityCallable(
+                                        $access_helper->isUserLoggedIn($access_helper->isRepositoryReadable(
+                                            static function () use ($dic): bool {
+                                                return true;
```
Is this correct? If yes please remove the dead code.

fhelfer committed:
license changes dashboard

fhelfer committed:
javascript dashboard administration

fhelfer committed:
dashboard block gui review

missed some

dashboard review

some follow up fixes + return types

fhelfer committed:
dashboard settings bugfix ux sortation

fhelfer committed:
dashboard ux bugfixes studyprogramme & learning sequences

fhelfer committed:
dashboard change order of blocks

fhelfer committed:
fix order dashboard view constants

fhelfer committed:
bugfix migrate missing settings in dashboard

fhelfer committed:
dashboard feault all sortations active & disable inactive studyprogrammes on dashboard

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,79 @@ public function editSettings(): void
+            )
+            ->withAdditionalOnLoadCode(
+                static fn (string $id) =>
+                    "$id.setAttribute('data-checkbox', 'activeSorting$view');
```
Please use `document.querySelector('#$id')` or `document.getElementById('$id')`.
All ID's are available as global variables but IMO there are 3 issues with using them:
- If the implementation of `withAdditionalOnLoadCode` changes and $id contains `-` (minus) this code will break (will be replaced with `my-id.setAttribute` which is the same as `my - id.setAttribute`).
- You can shadow them with new global variables which will give you the wrong element (or error).
```html
<div id="hej"></div>
```
```js
console.log(hej) // <div id="hej">
var hej = 3;
console.log(hej) // 3
```
- People don't expect this. I don't think that most people know that all ID's are automatically global variables and will have a harder time to understand this code (they will probably search for the place where this variable is defined as in most placed people use `$()`, `document.querySelector` or `document.getElementById` to access elements by ID.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,79 @@ public function editSettings(): void
+            ->withAdditionalOnLoadCode(
+                static fn (string $id) =>
+                    "$id.setAttribute('data-checkbox', 'activeSorting$view');
+                    document.addEventListener('DOMContentLoaded', function () {
```
I'm pretty sure that this code will never be executed.
You are already in a _withAdditional**OnLoad**Code_ block, so you can be sure that the content is already loaded. The `DOMContentLoaded` event was already triggered.

fhelfer replied:
It breaks when i remove it,
It kind of ensures that both ids have been set in data tags

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,79 @@ public function editSettings(): void
+            ->withRequired(false)
+            ->withAdditionalOnLoadCode(
+                static fn (string $id) =>
+                    "$id.setAttribute('data-select', 'sorting$view');"
```
Same as the first comment.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,89 @@
+ * @param {int} view
+ */
+function handleUserInputForSortationsByView(view) {
+    const selectionInput = document.querySelectorAll('[data-select="sorting' + view + '"]')[0];
```
Please use `querySelector(...)` instead of `querySelectorAll(...)[0]`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,89 @@
+ */
+function handleUserInputForSortationsByView(view) {
+    const selectionInput = document.querySelectorAll('[data-select="sorting' + view + '"]')[0];
+    const checkboxInput = document.querySelectorAll('[data-checkbox="activeSorting' + view + '"]')[0];
```
Please use `querySelector(...)` instead of `querySelectorAll(...)[0]`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,89 @@
+    const checkboxInput = document.querySelectorAll('[data-checkbox="activeSorting' + view + '"]')[0];
+    disableMinusOption(selectionInput);
+    updateSelectionInputWithActivatedOptions(selectionInput, checkboxInput);
+    const selectedOption = selectionInput.querySelectorAll('option[selected=\"selected\"]')[0];
```
Please use `querySelector(...)` instead of `querySelectorAll(...)[0]`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,89 @@
+function updateSelectionInputWithActivatedOptions(selectionInput, checkboxInput) {
+    selectionInput.querySelectorAll('option').forEach(function (element) {
+        const checkbox = checkboxInput.querySelector('input[value=\"' + element.value + '\"]')
+        if (checkbox &&  checkbox.checked === false) {
```
Please check if the elements you select are indeed optional. In this file most `null` `querySelector(...)` calls are silenced with an `if`. This should only be done if the node is indeed optional and not to prevent an error. It is ok to **let JS throw an error** if something happens that is vital for your code. This makes it easier to debug when something doesn't work.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,89 @@
+    } else {
+        checkboxInput.querySelectorAll('input[type=\"checkbox\"]:disabled').forEach(function (element) {
+            element.removeAttribute('disabled');
+            }
```
Close brace doesn't align with the open brace line. Please run a linter or something over this file. There are missing semicolons (either remove all semicolons or add all), mismatch of open & close indentation and superfluous white spaces.

lscharmer commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,168 @@
+        $relevant = false;
+
+        $ls_lp_items = $obj->getLSLearnerItems($this->user->getId());
+        if (count($ls_lp_items) === 0) {
```
Please use `$ls_lp_items` === []` instead`

lscharmer replied:
Please use `$ls_lp_items === []` instead.

lscharmer commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,168 @@
+
+        foreach ($ls_lp_items as $item) {
+            if ($item->getLearningProgressStatus() === ilLPStatus::LP_STATUS_IN_PROGRESS_NUM) {
+                $relevant = true;
```
Please use `return true` instead, as it is never set to false again.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -91,232 +72,389 @@ public function __construct()
+            foreach ($group as $datum) {
+                $cards[] = $this->getCardForData($datum);
+            }
+            $cards = array_filter($cards, static function ($card) {
```
Please use `array_filter($cards)` instead.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -91,232 +72,389 @@ public function __construct()
+                $data = array_merge(...array_values($data));
+                $data = $this->sortByTitle($data);
+                return ['' => $data];
+                break;
```
Dead code. Please remove.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -331,11 +469,54 @@ public function manageObject(): string
+                    $ui->factory()->link()->standard($item["txt"], $item["url"]),
+                $groupedCommands[0]
+            );
+            if (count($actions) > 0) {
```
PLease use `$actions !== []` instead.

lscharmer commented on Services/Dashboard/Block/classes/class.ilStudyProgrammeBlockGUI.php:
```diff
@@ -18,48 +16,36 @@
         foreach ($rows as $row) {
             $prg = ilObjStudyProgramme::getInstanceByObjId($row->getNodeId());
-            if (! $this->isReadable($prg)) {
+            if (! $this->isReadable($prg) || !$prg->isActive()) {
```
Code syle. Please remove the space between the `!` and `$this->isReadab...`

lscharmer commented on Services/Dashboard/classes/class.ilDashboardGUI.php:
```diff
@@ -147,33 +153,21 @@ public function executeCommand(): void
+                if (strtolower($cmdClass = $this->ctrl->getCmdClass()) === strtolower(ilSelectedItemsBlockGUI::class)) {
+                    $gui = new $cmdClass();
+                    $ret = $this->ctrl->forwardCommand($gui);
+                    if ($ret != "") {
```
Please use either `!$ret` or `$ret !== ""` depending on the fact that you know the type of `$ret` or not.

lscharmer commented on Services/Dashboard/classes/class.ilDashboardGUI.php:
```diff
@@ -219,14 +213,17 @@ public function executeCommand(): void
+            case strtolower(ilStudyProgrammeBlockGUI::class):
+                $gui = new $next_class();
+                $ret = $this->ctrl->forwardCommand($gui);
+                if ($ret != "") {
```
Please use either `!$ret` or `$ret !== ""` depending on the fact that you know the type of `$ret` or not.

fhelfer committed:
dashboard ui fixes

fhelfer committed:
dashboard standarize ListItem/Card HMTL

fhelfer committed:
dashboard codestyle fixes

js lint

fhelfer committed:
itemgui commands studyprogramme dashboard

fhelfer committed:
remove items from blocks dashboard bugfix

fhelfer committed:
wip roundtrip modal dashboard

fhelfer committed:
dashboard reoundtrip modal hacky content replace

iszmais commented:
@fhelfer Nur ein Reminder das der PR vor dem Merge in den Core gesquashed wird. 
Persönlich finde ich viele Commits mit detailreichen Messages natürlich super, aber wenn du dir die Mühe nur für den Kern machst kannst du es dir auch sparen :smile:

fhelfer committed:
refactoring remove_multiple items from dashboard

fhelfer committed:
removed unused dashboard lng

fhelfer committed:
disable minus option removed

fhelfer committed:
add CustomCommands with integer refid on dashboard

fhelfer committed:
bugfix read wrong settings keyword from def dash sorting

fhelfer committed:
remove modal footer from dashboard modal

fhelfer committed:
dashboard modal top button text

fhelfer commented:
@iszmais & @lscharmer, we changed a few more things maybe you can review again
for reference: not approved changes begin after https://github.com/ILIAS-eLearning/ILIAS/pull/4941/commits/cb2ce0502ff687fb21a4abbeffababd9db24c888
we added a roundtrip modal for removal of items from dashboard. Because the modal cannot handle asynchronous POST Requests we had to replace the content with javascript.

iszmais commented on Modules/StudyProgramme/classes/class.ilStudyProgrammeDIC.php:
```diff
@@ -340,17 +340,6 @@ protected static function buildDIC(): Container
             );
         };
 
-        $dic['ilStudyProgrammeDashboardViewGUI'] = static fn ($dic) =>
```
Im not sure if this violates a conventions, but im **personaly** not a fan to shorten semantic relevant keywords.
The time where the minification of code in this scope had a relevance are long ago. Thefore this just decrease the readability imo.

@mjansenDatabay are you aware of any convention in ilias for this?


Otherwise this would just be a recomandation rather than a change request.

iszmais commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -915,6 +917,11 @@ public function getHTMLNew(): string
                 });
             }
+
+            if (isset($command['modal']) && $command['modal'] instanceof \ILIAS\UI\Component\Modal\Modal) {
```
Pls put that namespace inside a use statement

(Same for all other edited lines, exepcetions can be made if the context where the changes where made follows a different approach)

And keep care not to mash up interrfaces and implementations :smile:

lscharmer replied:
+1

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+        $this->data = array_filter($a_data);
+    }
+
+    public function groupItemsByStartDate(): array
```
Add a default return decription inside the docs. Either something like `<string, int>` or, if this is not definable, a descriptive sentence.

Same for other array return or parameters

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+    public function changePDItemPresentationObject(): void
+    {
+        $this->viewSettings->storeActorPresentationMode(
+            \ilUtil::stripSlashes((string) ($this->http->request()->getQueryParams()['presentation'] ?? ''))
```
Remove \

iszmais commented on Services/Dashboard/ItemsBlock/classes/class.ilPDSelectedItemsBlockViewSettings.php:
```diff
@@ -388,13 +408,83 @@ public function getCurrentView(): int
+
+    public function enableRecommendedContent(bool $status): void
+    {
+        $this->settings->set('enable_recommended_content', (string) !$status);
```
We want to prevent casting when we actualy mean to translating something.
Strictly said there is no string value of a bollean (or at least there shouldnt be). Its a feature that PHP offers natively wich is outsourced it translation, parser or transmutation classes in type safe Languages.

Also i would be carefull with "semantic" casts in general. This often leave a wild field of interpretations that are often miseinterpred. In native PHP true is "1" and false is "" while in most settings false can also be "0". Sometimes even "false" will be interpred as false. Im not saying thats good, but we have to deal with it...

Therefore i would appreciate if you clearify this state (and the 2 following) by more than a native cast e.g.
```($status) ? "1" : "0"```;

iszmais commented on Services/Dashboard/templates/default/tpl.remove_multiple_modal_id_wrapper.html:
```diff
@@ -0,0 +1,3 @@
+<div id="pd_unsubscribe_multiple">
+    {CONTENT}
+</div>
```
Add an EOL

iszmais reviewed (changes requested):
Mostly looks good to me, just some more or less relevant conventional changes.

Note that i just overflew the JS parts, i leave those in the trustfully hand of our executive Java lead developer @lscharmer .

Due to a missing functionality in Github i wont approve this request when all changes are applied. I will run a testing phase after all review are approved and give my approval when this pahse was finished successsfully.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+    }
+    updateSelectionInputWithActivatedOptions(selectionInput, checkboxInput);
+    ensureLastOptionNotDeselectable(checkboxInput);
+    checkboxInput.onchange = function () {
```
Plase use:
```js
checkboxInput.addEventListener('change', function () {
```
instead of `x.onchange = ...`

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+            if (checkbox.checked === false) {
+                element.setAttribute('disabled', 'disabled');
+            }
+            if (checkbox.checked === true) {
```
Please use an `else` clause instead of the second `if`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+/**
+ * @param {int} view
+ */
+function handleUserInputForSortationsByView(view) {
```
This seems to be the only "public" function. Please put everything into an own scope and only make this function available in the global scope like this:
```js
const handleUserInputForSortationsByView = (function () {
    .... everything else like function updateSelectedValueToFirstPossibleOption(selectionInput) { ... } etc
    return function (view) {
    ...
    };
})();
```

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+/**
```
Please rename this file. The `il` prefix is unnecessary as this has nothing to do with namespacing (and even in PHP one should only use this where normal namespaces cannot be used).
Please also remove the `Dashboard` prefix as we know that we are in the dashboard already.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+function updateSelectionInputWithActivatedOptions(selectionInput, checkboxInput) {
+    selectionInput.querySelectorAll('option').forEach(function (element) {
+        if (element.value !== '') {
+            const checkbox = checkboxInput.querySelector('input[value=\"' + element.value + '\"]');
```
No need to escape a double quote `"` in a single quoted string `'`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+ */
+function switchBackToDefaultIfEnabled(selectionInput, checkboxInput) {
+    selectionInput.querySelectorAll('option').forEach(function (element) {
+        const checkboxOption = checkboxInput.querySelector('input[value=\"' + element.value + '\"]');
```
No need to escape a double quote `"` in a single quoted string `'`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+ */
+function updateSelectedValueToFirstPossibleOption(selectionInput) {
+    const selectedDefaultSortation = selectionInput.querySelector('option[selected="selected"]');
+    if (!selectedDefaultSortation || selectedDefaultSortation.getAttribute('disabled') === 'disabled') {
```
Can you combine the `disabled` attribute into the selector? Like this:
```js
const selectedDefaultSortation = selectionInput.querySelector('option[selected="selected"]:not([disabled="disabled"])');
```
Or is it important that you select the first `selected` input and then check for `disabled`?

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+ * @param {Element} checkboxInput
+ */
+function ensureLastOptionNotDeselectable(checkboxInput) {
+    const checkedCheckbox = checkboxInput.querySelectorAll('input[type=\"checkbox\"]:checked');
```
No need to escape a double quote `"` in a single quoted string `'`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+    if (checkedCheckbox.length === 1) {
+        checkedCheckbox[0].setAttribute('disabled', 'disabled');
+    } else {
+        checkboxInput.querySelectorAll('input[type=\"checkbox\"]:disabled').forEach(function (element) {
```
No need to escape a double quote `"` in a single quoted string `'`.

lscharmer commented on Services/Dashboard/Administration/js/ilDashboardSortationUserInputHandler.js:
```diff
@@ -0,0 +1,82 @@
+function handleUserInputForSortationsByView(view) {
+    const selectionInput = document.querySelector('[data-select="sorting' + view + '"]');
+    const checkboxInput = document.querySelector('[data-checkbox="activeSorting' + view + '"]');
+    const selectedOption = selectionInput.querySelector('option[selected=\"selected\"]');
```
No need to escape a double quote `"` in a single quoted string `'`.

lscharmer commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -171,19 +171,21 @@ public function getTitle(): string
     }
 
-    public function correctOffset(): void
+    public function correctOffset(int $a_offset): bool
```
Please remove the `a_` suffix from the parameter `$a_offset`.

lscharmer commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -314,12 +316,13 @@ public function getRowTemplateDir(): string
     }
 
-    public function addBlockCommand(string $a_href, string $a_text, string $a_onclick = ""): void
+    public function addBlockCommand(string $a_href, string $a_text, string $a_onclick = "", ILIAS\UI\Component\Modal\RoundTrip $modal = null): void
```
Please `use` the `RoundTrip` at the beginning of the file, so you can just use `RoundTrip` here.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,79 @@ public function editSettings(): void
+                $this->viewSettings->getPresentationViews()
+            )
+        );
+        return $form;
```
Dead code. Please remove.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -118,6 +122,79 @@ public function editSettings(): void
+                        handleUserInputForSortationsByView($view);
+                    });"
+            );
+        $options = [];
```
Duplicated code. The `$options` are already filled in line `155` with the same value as with this `foreach`.

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -160,14 +247,29 @@ public function initForm(): \ILIAS\UI\Component\Input\Container\Form\Standard
+    {
+        return $this->ui_factory->input()->container()->form()->standard(
+            $this->ctrl->getFormAction($this, 'savePresentation'),
+            array_map(
```
Duplicated `array_map(...)`. Method `initSortingForm` line `138`. Please extract this into an own method. The only difference is `$this->getViewSorting` vs. `$this->getViewPresentation` which can be supplied as parameters.

fhelfer committed:
review dashboard #1

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +308,107 @@ public function setSettingsSubTabs(string $a_active): void
-        $this->tpl->setOnScreenMessage('success', $lng->txt("msg_obj_modified"), true);
-        $ctrl->redirect($this, $redirect_cmd);
+        foreach ($form_data as $view => $view_data) {
+            if (!isset($view_data['avail_sorting']) || !is_array($view_data['avail_sorting'])) {
```
Can be shortened to `!is_array($view_data['avail_sorting'] ?? null)`

lscharmer commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +308,107 @@ public function setSettingsSubTabs(string $a_active): void
-        $ctrl->redirect($this, $redirect_cmd);
+        foreach ($form_data as $view => $view_data) {
+            if (!isset($view_data['avail_sorting']) || !is_array($view_data['avail_sorting'])) {
+                $view_data['avail_sorting'] = [$view_data['default_sorting']];
```
I think the key `default_sorting` needs to be checked for existence, as it is checked in line 403 too (and therefore optional).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+    {
+        $groupedCards = [];
+        foreach ($this->loadData() as $title => $group) {
+            $cards = [];
```
Can be shortened to `$cards = array_map($this->getCardForData(...), $group);`

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+        $this->viewSettings->parse();
+        $this->initData();
+
+        if ($this->viewSettings->isTilePresentation()) {
```
Duplicated code. in constructor line: 69. Please extract this into an own method.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+            $grouped_items[$title][] = $item;
+        }
+        ksort($grouped_items);
+        foreach ($grouped_items as $key => $group) {
```
Can be shortened to `$grouped_items = array_map($this->sortByTitle(...), $grouped_items);

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+        }
+
+        $presentations = $this->viewSettings->getSelectablePresentationModes();
+        if (count($presentations) > 1) {
```
Unnecessary `if`. Please remove it.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+        $button = $this->ui->factory()->button()->standard($this->getRemoveMultipleActionText(), '#')
+        ->withOnLoadCode(function ($id) use ($url) {
+            return "
+            $('#$id').on('click', function() {
```
Please move this into an own file. Replace var with const / let. Consider replacing jQuery calls to native calls if there is no obvious benefit to keeping jQuery (I will probably do a proper review when it has an own file.

fhelfer replied:
moved to own file and removed native calls

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,782 @@
+        return $item_list_gui;
+    }
+
+    public function sortByTitle(array $data, bool $asc = true): array
```
This method should be `private`

fhelfer committed:
review dashboard #2

fhelfer committed:
review dashboard #3

fhelfer committed:
review dashboard #4 blockdto

fhelfer committed:
review dashboard #4 blockdto

fhelfer committed:
typo

fhelfer committed:
short path

fhelfer committed:
review dashboard #5

fhelfer committed:
review dashboard #6

fhelfer committed:
fix copyright

iszmais commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -171,19 +173,21 @@ public function getTitle(): string
+        if ($this->correctOffset($a_offset)) {
+            $this->offset = $a_offset;
+        } else {
+            throw new ilException("ilBlockGUI::setOffset(): Offset out of range.");
```
Very picky, but...
To be out of range you need a range to be checked, therefore 2 paramters.
Since the check is missing the lower limit its more lika a "max exceeded " exception.

You may solves this by rephrasing. Otherwise see me recommendation on the function itsself.

iszmais commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -171,19 +173,21 @@ public function getTitle(): string
     }
 
-    public function correctOffset(): void
+    public function correctOffset(int $offset): bool
```
If we have a check here i would prefere  a full range check (> and <) if possible (s. a.).

Also the function name is a bit irritating, since it provides already a statement,therfore i would expcept the function ittself to throw an error if this statement is wrong. Due to that ILIAS has a soft convention omn the preffix `check` instead of `correct`

```php
function checkThat() //Checks that and return a bool or someting wich implies the check succeedes or failed
```

```php
function correctThat() //Says That is already correct, therefore must throw an error if not since the functions precondition is wrong.
//Alternatively this could also be a proverb wich states the function corrects the value of  That, manipulating it or returning a transformed correct value.
//Neither is the case here
```

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -114,11 +123,86 @@ public function editSettings(): void
+    {
+        $this->tabs_gui->activateTab("settings");
+        $this->setSettingsSubTabs("sorting");
+        $ui = $this->ui;
```
Kinda an overkill if its only used 1 time in a function.

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -114,11 +123,86 @@ public function editSettings(): void
+                return $this->ui_factory->input()->container()->form()->standard(
+                    $this->ctrl->getFormAction($this, 'save' . $mode),
+                    array_map(
+                        fn (int $view): Section =>
```
I appreciate the usage of new PHP functionalities, but recommend the usage with care since this tend to inherit issues to future usages.
Therefore I encurage you to look over your arrow functions critical and ensure that their requirement is necessary.
Further I would appreciate to establish a consistent code style here wich rest on the default function style like

```php
fn (int $that): This => $this->doStuff(
    $verylongparametera,
    $verylongparameterb
)

//or

fn (int $that): This => (
    $this->doVeryComlicatedStuffWIthAVerLongName(
        $a,
        $b
    )
)

//or

fn (int $that): This => $this->doStuff($a, $b)
``` 
 Or feel free to propse an own style if you have a preferation.

lscharmer commented on Services/Dashboard/Administration/js/SortationUserInputHandler.js:
```diff
@@ -0,0 +1,84 @@
+    function switchBackToDefaultIfEnabled(selectionInput, checkboxInput) {
+        selectionInput.querySelectorAll('option').forEach(function (element) {
+            const checkboxOption = checkboxInput.querySelector('input[value="' + element.value + '"]');
+            if (element.hasAttribute('default') && checkboxOption && checkboxOption.checked === true) {
```
```suggestion
            if (element.hasAttribute('default') && checkboxOption && checkboxOption.checked) {
```

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -114,11 +123,86 @@ public function editSettings(): void
+    public function getSettingsForm(): StandardForm
     {
         $ui = $this->ui;
         $f = $ui->factory();
```
Im fine with aliases but but they still should keep some semantical context.
Single Char variables are mostly reserved for loop indexes and math operations ($f would be a great variable to store a linear function in or some Fahrenheit temperature).

Maybe as a general rule you could take that ouy always look on the right side of an alias to improve its length, than on the left.
In this special case for example you could alias not only the factory, but for the most cases the input->field ,too:
```
$field = $this->ui->factory()->input()->field()
```

Further the usage of aliases should provide a significant reduction of code.
The alias `$ui` has such less occurences that it actualy inflates the code :smile: 
I personally use "normal" aliases only for > 3 Occurences.

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +317,107 @@ public function setSettingsSubTabs(string $a_active): void
+            $this->viewSettings->storeViewPresentation(
+                $view,
+                $view_data['default_pres'],
+                $view_data['avail_pres'] ?: []
```
Do you want to proof `== false` or `=== null` here?
If second, use `??` since its more strict.

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +317,107 @@ public function setSettingsSubTabs(string $a_active): void
-        $this->viewSettings->storeViewSorting(
-            $view,
-            $form_data['sortation']['default_sort'],
-            $form_data['sortation']['avail_sort'] ?: []
```
Do you want to proof `== false` or `=== null` here?
If second, use `??` since its more strict.

iszmais commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -204,153 +317,107 @@ public function setSettingsSubTabs(string $a_active): void
-        $this->viewSettings->storeViewPresentation(
-            $view,
-            $form_data['presentation']['default_pres'],
-            $form_data['presentation']['avail_pres'] ?: []
```
Do you want to proof `== false` or `=== null` here?
If second, use `??` since its more strict.

lscharmer commented on Services/Dashboard/Administration/js/SortationUserInputHandler.js:
```diff
@@ -0,0 +1,84 @@
+/**
+ * @param {int} view
```
```suggestion
 * @type {(view: int) => void}
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilBlockDTO.php:
```diff
@@ -0,0 +1,165 @@
+    private ?ilDateTime $endDate;
+    private array $additional_data;
+
+    public function __construct(
```
Please use Constructor pomotion: https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion

lscharmer commented on Services/Dashboard/Block/classes/class.ilBlockDTO.php:
```diff
@@ -0,0 +1,165 @@
+
+declare(strict_types=1);
+
+class ilBlockDTO
```
Please use a namespace and drop the `il` prefix.

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+        }
+    }
+
+    #[NoReturn]
```
Feel free to use `NoReturn`.
However since that function end with an exit you may better use the [`never`](https://www.php.net/manual/de/language.types.never.php) return type, wich implies NoReturn.

But i encourage you to add #[NoReturn] to (prolly most of) your other functions wich return void since its a heavy impovement for developing within IDEs.

iszmais replied:
Or, if you can, remove the exit like in https://github.com/ILIAS-eLearning/ILIAS/pull/4941#discussion_r1182484724

lscharmer replied:
The `NoReturn` / `never` is still valid, as the `$DIC->http()->close()` is effectively an `exit` (but a testable exit).

iszmais replied:
I agree but would not recommend to set `never` until the returntype of `close()` was changed, too.
Otherwise we might encounter problems with the function chain.

But `NoReturn` should still be possible.

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+            'removeFromDeskRoundtrip',
+            '',
+            true
+        ) . '&page=confirm';
```
Use `ctrl->setParameter()` instead of string concatination, if possible.

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+        return (new ilDashObjectsTableRenderer($this))->render($grouped_items);
+    }
+
+
```
Remove line

iszmais commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,154 @@
+use ILIAS\UI\Component\Symbol\Icon\Standard;
+
+/**
+ *
```
Remove line

iszmais commented on Services/Dashboard/Block/classes/class.ilSelectedItemsBlockGUI.php:
```diff
@@ -0,0 +1,133 @@
+    public function emptyHandling(): string
+    {
+        $this->lng->loadLanguageModule('rep');
+        $txt = $this->lng->txt("rep_fav_intro1") . "<br>";
```
You can also change the language vars directly, if you are the only one using them.
Language variables support minor html like `<br>` and stuff.

iszmais commented on Services/Dashboard/Block/classes/class.ilSelectedItemsBlockGUI.php:
```diff
@@ -0,0 +1,133 @@
+        $txt .= sprintf(
+            $this->lng->txt('rep_fav_intro2'),
+            $this->getRepositoryTitle()
+        ) . "<br>";
```
s.a.

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+        $this->init();
+        if ($this->ctrl->isAsynch()) {
+            echo $this->getHTML();
+            exit;
```
Please use `$DIC->http()` to send a response and `exit` (with `$DIC->http()->close()` etc).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+            return strcmp($left->getTitle(), $right->getTitle());
+        };
+
+        uasort($groups['upcoming'], static fn (ilBlockDTO $left, ilBlockDTO $right): int => $orderByDate($left, $right));
```
```suggestion
        uasort($groups['upcoming'], $orderByDate);
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+
+        uasort($groups['upcoming'], static fn (ilBlockDTO $left, ilBlockDTO $right): int => $orderByDate($left, $right));
+        uasort($groups['ongoing'], static fn (ilBlockDTO $left, ilBlockDTO $right): int => $orderByDate($left, $right, false));
+        uasort($groups['ended'], static fn (ilBlockDTO $left, ilBlockDTO $right): int => $orderByDate($left, $right));
```
```suggestion
        uasort($groups['ended'], $orderByDate);
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+                $modal = $this->ui->factory()->legacy($this->confirmRemoveObject());
+        }
+        echo $this->ui->renderer()->renderAsync($modal);
+        exit;
```
Please use `$DIC->http()` to send a response and exit (with `$DIC->http()->close()` etc).

lscharmer commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -0,0 +1,683 @@
+        $this->ctrl->setParameter($this, 'view', $this->viewSettings->getCurrentView());
+
+        $refIds = (array) ($this->http->request()->getParsedBody()['id'] ?? []);
+        if (0 === count($refIds)) {
```
```suggestion
        if ([] === $refIds) {
```

iszmais reviewed (changes requested):
LGTM :+1: 
just some code style requests and recommandations.
Feel free to contact me if you feel unsure about specific points.

Greetings,
@iszmais

lscharmer commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,154 @@
+
+    protected function isRelevantLso(ilObjLearningSequence $obj): bool
+    {
+        $relevant = false;
```
I would just `return false` instead of `return $relevant`.

lscharmer commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,154 @@
+
+    public function addCustomCommandsToActionMenu(ilObjectListGUI $itemListGui, int $ref_id): void
+    {
+        return;
```
Unnecessary statement. Please remove.

lscharmer commented on Services/Dashboard/Block/classes/class.ilLearningSequenceBlockGUI.php:
```diff
@@ -0,0 +1,154 @@
+    public function confirmedRemoveObject(): void
+    {
+        $refIds = (array) ($this->http->request()->getParsedBody()['ref_id'] ?? []);
+        if (0 === count($refIds)) {
```
```suggestion
        if ([] === $refIds) {
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilMembershipBlockGUI.php:
```diff
@@ -0,0 +1,129 @@
+
+    public function addCustomCommandsToActionMenu(ilObjectListGUI $itemListGui, int $ref_id): void
+    {
+        return;
```
Unnecessary statement. Please remove.

lscharmer commented on Services/Dashboard/Block/classes/class.ilMembershipBlockGUI.php:
```diff
@@ -0,0 +1,129 @@
+    public function confirmedRemoveObject(): void
+    {
+        $refIds = (array) ($this->http->request()->getParsedBody()['ref_id'] ?? []);
+        if (0 === count($refIds)) {
```
```suggestion
        if ([] === $refIds) {
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilRecommendedContentBlockGUI.php:
```diff
@@ -0,0 +1,151 @@
+    {
+        $rec_manager = new ilRecommendedContentManager();
+        $refIds = (array) ($this->http->request()->getParsedBody()['ref_id'] ?? []);
+        if (0 === count($refIds)) {
```
```suggestion
        if ([] === $refIds) {
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilSelectedItemsBlockGUI.php:
```diff
@@ -0,0 +1,133 @@
+
+    public function addCustomCommandsToActionMenu(ilObjectListGUI $itemListGui, int $ref_id): void
+    {
+        return;
```
Unnecessary statement. Please remove.

lscharmer commented on Services/Dashboard/Block/classes/class.ilSelectedItemsBlockGUI.php:
```diff
@@ -0,0 +1,133 @@
+    public function confirmedRemoveObject(): void
+    {
+        $refIds = (array) ($this->http->request()->getParsedBody()['ref_id'] ?? []);
+        if (0 === count($refIds)) {
```
```suggestion
        if ([] === $refIds) {
```

lscharmer commented on Services/Dashboard/Block/classes/class.ilStudyProgrammeBlockGUI.php:
```diff
@@ -0,0 +1,219 @@
+
+    public function addCustomCommandsToActionMenu(ilObjectListGUI $itemListGui, int $ref_id): void
+    {
+        return;
```
Unnecessary statement. Please remove.

lscharmer commented on Services/Dashboard/Block/classes/class.ilStudyProgrammeBlockGUI.php:
```diff
@@ -0,0 +1,219 @@
+
+    public function confirmedRemoveObject(): void
+    {
+        return;
```
Unnecessary statement. Please remove.

lscharmer commented on Services/Dashboard/Block/js/replaceModalContent.js:
```diff
@@ -0,0 +1,27 @@
+ * @param {int} view
+ * @param {string} url
+ */
+const replaceModalContent = (function (id, view, url) {
```
```suggestion
const replaceModalContent = function (id, view, url) {
```

lscharmer commented on Services/Dashboard/Block/js/replaceModalContent.js:
```diff
@@ -0,0 +1,27 @@
+        modalFooter.parentNode.removeChild(modalFooter);
+        il.Util.ajaxReplacePostRequestInner(url, post_data, 'pd_unsubscribe_multiple_view_' + view);
+    });
+});
```
Missing newline at the end of file.

lscharmer commented on Services/Dashboard/Block/js/replaceModalContent.js:
```diff
@@ -0,0 +1,27 @@
+        }
+        const modal = document.querySelector('div[data-modal-name="remove_modal_view_' + view + '"]');
+        let post_data = '';
+        for (let i = 0; i < selected_ids.length; i++) {
```
This for loop and the `selected_ids` variable are unnecessary. You can just create the post data in the previous loop in line `12`.

lscharmer commented on Services/JavaScript/js/Basic.js:
```diff
@@ -171,6 +181,23 @@ il.Util = {
+	 */
+	sendAsyncAjaxPostRequestToUrl: function(url, data, args, succ_cb) {
+		var cb =
+			{
```
Please start the object in the `var cb = ` line and indent this nicely.

fhelfer committed:
review dashboard #7

add remove to #7

add use to #7

lscharmer reviewed (changes requested):
Hi @fhelfer,

- [x] Please change the `echo` usages before `$DIC->http()->close()` to use `$DIC->http()->sendResponse(...)` etc. to send the HTML content.
- [x] Drop the `class.` prefix in the filename of `class.BlockDTO.php`

Besides that the PR looks really good, these should be that last change requests.

Best regards
@lscharmer

iszmais commented:
I agree to @lscharmer ,
There is 1 thread left from my sight https://github.com/ILIAS-eLearning/ILIAS/pull/4941#discussion_r1082653417
Otherwise this is fine.

Thx,
@iszmais

fhelfer committed:
review dashboard #8

iszmais commented:
@fhelfer Something generic:
I would recommend you not to use `#` within your commit or version in general if you use github or gitlab.
Reason ist that both tool try to resolve `#` as an referer to another PR, commit or is (just like #4941 refers to your PR).
This not only links your commits unintetionally to other unrelated ticket, it also leaves a mention within that ticket that it was refered somewhere else.

It has no effect on the progress itself but may lead to a lot of confusion :smile:

iszmais commented:
@fhelfer Sometinhg generic:
I would recommend you not to use `#` within your commit or version in general if you use github or gitlab.
Reason ist that that both tool try to resolve `#` as an referer to another PR, commit or is (just like #4941 refers to your PR).
This not only links your commits unintetionally to other unrelated ticket, it also leaves a mention within that ticket that it was refered somewhere else.

It has no effect on the progress itself but may lead to a lot of confusion :smile:

iszmais reviewed (approved):
LGTM :+1: 

I will await the response of Mjansen and then test it before merging.

Greetings,
@iszmais

iszmais reviewed (approved):
LGTM :+1: 

I will await the response of @mjansenDatabay  and then test it before merging.

Greetings,
@iszmais

mjansenDatabay commented:
> LGTM 👍
> 
> I will await the response of @mjansenDatabay and then test it before merging.
> 
> Greetings, @iszmais

After internal approval: Please add the component maintainers for the folders you changed files in.

mjansenDatabay commented on Services/Mail/classes/class.ilPDMailBlockGUI.php:
```diff
@@ -98,6 +98,7 @@ public function getHTML(): string
 
         $this->getMails();
         $this->setData($this->mails);
+        $this->loadData();
```
Why is this change necessary? Why should block implementations be responsible to call this?

mjansenDatabay commented on Services/Membership/classes/class.ilMembershipOverviewGUI.php:
```diff
@@ -46,9 +45,9 @@ public function executeCommand(): void
         $this->main_tpl->setTitle($this->lng->txt("my_courses_groups"));
         switch ($next_class) {
-            case "ilpdmembershipblockgui":
+            case "ilmembershipblockgui":
```
I suggest to use `strtolower(ilMembershipBlockGUI::class)` here

mjansenDatabay commented on Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php:
```diff
@@ -47,38 +47,38 @@ public function __construct(?ilObjUser $user = null)
-            $ctrl->setParameterByClass("ilPDSelectedItemsBlockGUI", "view", "0");
-            $ctrl->setParameterByClass("ilPDSelectedItemsBlockGUI", "col_side", "center");
-            $ctrl->setParameterByClass("ilPDSelectedItemsBlockGUI", "block_type", "pditems");
+            $ctrl->setParameterByClass("ilSelectedItemsBlockGUI", "view", "0");
+            $ctrl->setParameterByClass("ilSelectedItemsBlockGUI", "col_side", "center");
+            $ctrl->setParameterByClass("ilSelectedItemsBlockGUI", "block_type", "pditems");
```
I suggest to use the class constant instead

mjansenDatabay replied:
I suggest to use the class constants instead

mjansenDatabay commented on Services/Block/classes/class.ilBlockGUI.php:
```diff
@@ -33,7 +35,7 @@ abstract class ilBlockGUI
     public const PRES_MAIN_LIST = 3;		// main standard list panel
     public const PRES_MAIN_TILE = 4;		// main standard list panel
-    private int $offset;
+    private int $offset = 0;
```
I suggest to switch the approach if you have to modify all occurrences anyway. You should introduce a value object like `BlockPage` with a `PageSize` and a `PageOffset`. Some blocks might not support a pagination, so this should be `nullable`.

mjansenDatabay replied:
OK, my fault. I thought you introduced this `$offset` property.

mjansenDatabay commented on Services/Dashboard/Administration/classes/class.ilObjDashboardSettingsGUI.php:
```diff
@@ -24,10 +31,15 @@
+    public const VIEW_MODE_PRESENTATION = 'Presentation';
+    public const VIEW_MODE_SORTING = 'Sorting';
+    public const DASH_SORT_PREFIX = 'dash_sort_by_';
+    public const DASH_ENABLE_PREFIX = 'dash_enable_';
```
Please use a blank line to separate the constants from the properties

mjansenDatabay commented on Services/Dashboard/Administration/js/SortationUserInputHandler.js:
```diff
@@ -0,0 +1,84 @@
+/**
+ * @type {(view: int) => void}
+ */
+const handleUserInputForSortationsByView = (function() {
```
This constant polutes the global namespace, please check our JS guidelines, e.g.: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/docs/development/js

And: Why did you chose an upper case file name here, and a lower case filename in the other new JS file?

fhelfer replied:
```js
/* global $, il */
il.Dashboard = il.Dashboard || {};
/**
 * @type {(view: int) => void}
 */
 il.Dashboard.handleUserInputForSortationsByView = function() {
```

would this be the correct solution?

mjansenDatabay commented on Services/Dashboard/Block/js/replaceModalContent.js:
```diff
@@ -0,0 +1,23 @@
+ * @param {int} view
+ * @param {string} url
+ */
+const replaceModalContent = function (id, view, url) {
```
This constant polutes the global namespace, please check our JS guidelines, e.g.: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/docs/development/js 

And: Why did you chose a lower case file name here, and an upper case filename in the other new JS file?

mjansenDatabay commented on Services/Dashboard/classes/class.ilDashboardGUI.php:
```diff
@@ -220,14 +214,23 @@ public function executeCommand(): void
-                $gui = new ilStudyProgrammeDashboardViewGUI();
-                $this->ctrl->forwardCommand($gui);
+            case strtolower(ilObjStudyProgrammeGUI::class):
+                include_once("./Modules/StudyProgramme/classes/class.ilObjStudyProgrammeGUI.php");
```
Do we really need this `include_once` directive here? The autoloader should do the job.

fhelfer replied:
weirdly it does not, 
when accessing a study object via action menu from dashboard it crashes without this line

mjansenDatabay replied:
OK, but why? We should find the reasons and file an issue in Mantis for this problem. Maybe it is because the classname differs from the filename?

fhelfer committed:
review dashboard 9

fhelfer commented:
@mjansenDatabay can you assign me again, it seems i lost rights to push to the branch

iszmais commented:
> @mjansenDatabay can you assign me again, it seems i lost rights to push to the branch

By default the assigned is always the maintainer.
However you shoul still be able to push, the asignement does not interfere with the versioning at all. :sweat_smile: 
Sure thats the problem?

Greetings,
@iszmais

iszmais commented:
> @mjansenDatabay can you assign me again, it seems i lost rights to push to the branch

By default the assigned is always the maintainer.
However you shoul still be able to push, the asignement does not interfere with the versioning at all. :sweat_smile: 
Sure thats the problem?

Greetings,
@iszmais

fhelfer committed:
remove require studyprogramme

fhelfer committed:
remove js from global namespace

fhelfer committed:
fix standard listing panel heading no flex + loadData to set max_count ilBlockGUI

fhelfer committed:
remove laodData from MailBlocKGUI

fhelfer committed:
remove loadDatat from blockgui getHTML

mjansenDatabay commented:
@iszmais Please keep in mind that you **cannot** just merge this without the approval of all maintainers of the folders where files have been changed in.

iszmais commented:
@klees @alex40724 @smeyer-ilias i would appreciate if you could Review your depending Code.

A review of the scope of your responsible maintainance is sufficent, you dont have to review the whole PR.
Feel free to contanct @fhelfer for any conceptional issues you might see.

Could someone also add the maintainer of the `Object` component (fawinike)? Im not familiar with him and couldnt find a matching github account.

iszmais commented:
@fhelfer while testing i encountered some minor issues:
- Initially there are no set settings for the new areas. This problme resolve itself when i save every from in inside the dashboard settings once and doesnt occur afterward anymore. But there still should be an initial set of settings even without the manual save.
- Even with favourites activated i cant add to favourites within the dashboard (e.g. from my courses an groups) since the action menu entry seems missing. I stil can add to favourite through the menu inside the object itself. Im not sure if this is part of the feature (i wouldnt mind) but if not this should work.
- Over the modal "usubscribe mutliple" i can leave courses even if im the last administrator. This was already an issue in the past (https://github.com/ILIAS-eLearning/ILIAS/pull/4649) i assume this pr missed the fix.
- When i deactivate "my coruses and groups" the shown link in the dashboard settings send me to the main menue settings, but the title of the page is "dashboard"

Greetings,
@iszmais

klees reviewed (commented):
Hi @iszmais,

I'm not convinced that this is really the way to move forward with the integration of components into the Dashboard. If this is some intermediate step towards a bigger goal I'm fine with merging this, but I don't think that this should be the desired state.

Let me explain: Before this change, components (mostly) could integrate themselves into the Dashboard. There have been places, where the Dashboard needed to know the components that are integrated, but these only have been so many. Most of the code resided in the component that was integrated.

From my perspective, that would be the desired state: Components can integrate themselves in the Dashboard without the Dashboard knowing individual components that are integrated. This is good for various reasons:
* Components that want to integrate only need to know the Dashboard, while the Dashboard would need to know all components that are integrated. The further is simpler.
* Knowledge regarding the component stays in the component, the Dashboard does not need to know specific requirements for components when displaying them.
* When the Dashboard needs to know all integrated components it is closed for extension. Moving towards an "Adapatable Learning Environment" is one of five item in the [current strategy of the TB](https://docu.ilias.de/goto_docu_file_8005_download.html).

I do not stipulate that the current version of the Dashboard is there yet but this change certainly is a step in the exact opposite direction. I would appreciate if we move towards a solution where the Dashboard needs to know less (and at some point nothing) about the integrated components, not more.

If there is anything I can do to help you find such a solution, feel free to contact me, as always. If you already have something like this in mind, I would appreciate some rough sketch of the idea and agree to merge this.

Kind regards!

iszmais commented:
Thank you for your contribution @klees 

I totally understand your concern and partially agree with it.
After further investigation i think the best solution would be a dynamicly loaded setting for the Dashboard which sets the target option for any providing component.

But even if i approve the approach to make such central "gatherers" less knowing i think the dashboard might be a bit special to implement that pattern completely.
Imo the Dashboard has a special duty to a provide an intuitive user experience and therefore must provide some option for the adminitrator to guarantee such experience. As a introducting spage to an Ilias installation its viewed more ciritical in such point as e.g. the repository.
But i wouldnt see any reason why this both intention should contratdict.

However, im not quite sure what to do now since this is already and defined and approved Feature request. :smile: 
I would not put that extra load on said FR.
Imo this is, even if it might not be the perfect way, still an improvement to the current state and will make changes in said direction a lot easier in future.
Therefore i see no critical issue in merging this.

Greetings,
@iszmais

fhelfer committed:
favourites manager for my courses block

kergomard reviewed (commented):
Hi @fhelfer and sorry it passed me by that you added me as a reviewer!

Thank you very much for this PR and for implementing this features that makes the Dashboard more consistent!

...I think you already see the 'but' coming, right:
I cannot accept the change in `ilObjectListGUI`. Now, I know, this class is an abomination and I can understand what you are trying to do and why, but your implementation is going to brake very, very soon as we need to replace ilAdvancedSelectionListGUI. Additionally: Are you sure this also works with asynchronous loading of the actions? I simply don't know right now.

But now, the but of the but: Do you really need this? I will touch this class as soon as I can, but I'm not sure I'm going to make it for ILIAS 9. Sorry!

Best,
@kergomard

iszmais commented:
Thank you for the response @kergomard ,

Im not sure if i understand correctly. As far as i can see the changes made on that specific class arent critical and the return of the only added function is checked afterwards.
If you are referening to the specific return type i dont see any issue with chaning that on our side, but im still would recommend the current implementation since its based on the property type itsself.

Further i cant quite wrap my head around why this should be in issue due to your current ongoing changes? This seem to be a very standard prcedure in our way of versioning since, if you are planning to change the parameter in future this would imply to check all its direct occurences in that scope anyway. Or otherwise if your change preceed this PR we are boliged change the added function.
But that just seem to me like an normal PR resolvement as it normally occurs sometimes in parallel development.
And not as an blocking issue.


Lastly i cant really tell anything about the requirement of the change. For me it seems to be the way the StudyProgramme choose to present their list and is therefore not an issue of the dashboard itsself but carried on from it to the objectListGUI.
Due to that i assume you most likely will have the pointed out issues anyway within the communication with the SP since this PR doesnt really add something new.
But @fhelfer maybe can give a better overview on that front.

Greetings,
@iszmais

kergomard commented:
Hi @iszmais
I can put this very bluntly: You are exposing internals of the class. These internals are none of your business. Would it be intended for you to do anything with the commands, they would already be accessible. Every additional thing you expose of this class is an additional problem for future refactoring. For all intents and purposes: Consider this class final and deprecated (sadly I can't do the first part).
The only thing you can really expect from this class is either a rendered item or a rendered tile.

Best,
@kergomard

iszmais commented:
Thank you for the further explanation @kergomard !

I agree that that should be avoided.

@fhelfer can you take a further look at this if we have some alternatives here?

Greetings,
@iszmais

fhelfer committed:
remmove changes to objectlistgui for studyprogrammes

fhelfer commented:
Hello @kergomard,

thanks for the input. I removed the changes to OblectListGUI from this PR so the dashboard only uses the standard ItemListGUI::getAsListItem without custom changes.
But because the dashboard always displayed more informations for studyprogrammes i made a extra PR for the StudyprogrammeListGUI proposing adding these informations to display in dashboard and repository

Best,
@fhelfer

mjansenDatabay commented:
My two cents:

- Moving provider code from the specific components (`StudyProgramme`, `IndividualAssessment`, `RecommendedContent`) to a central folder (`Services/Dashboard/Block`) is a bad idea. This is a contradiction to the strategy of inverting the directions between the dashboard an its providers, which is stated at least for blocks in the `ROADMAP.md` https://github.com/ILIAS-eLearning/ILIAS/blob/release_8/Services/Dashboard/ROADMAP.md#move-to-global-screen-concepts. I am aware that the moved classes are acutally not considered `Blocks`, but if we try to adopt global screen concepts in the future, there must be some `Provider/Collector` concept based on interfaces even for this non-block classes.
- I agree that @iszmais must not do big steps forward to achieve the roadmap goals with **this** PR/change request, but of course we should not make the current situation even worse and move in the opposite direction.

fhelfer committed:
move dashboard blocks back to according folders

mjansenDatabay reviewed (changes requested):
I would even say: Don't rename the provider files (mentioned above) in this PR. It is hard (for me) to identify changes you made if you rename/move complete files.

fhelfer committed:
remove courses with last admin from dashbaord modal

fhelfer committed:
revert to old file names dashboard

fhelfer committed:
typo

fhelfer committed:
ilctrl structure

fhelfer committed:
add to favourites studyprogramme

fhelfer committed:
initial settings for dashboard sortation + mycourses link title

fhelfer commented:
> @fhelfer while testing i encountered some minor issues:
> 
> * Initially there are no set settings for the new areas. This problme resolve itself when i save every from in inside the dashboard settings once and doesnt occur afterward anymore. But there still should be an initial set of settings even without the manual save.
> * Even with favourites activated i cant add to favourites within the dashboard (e.g. from my courses an groups) since the action menu entry seems missing. I stil can add to favourite through the menu inside the object itself. Im not sure if this is part of the feature (i wouldnt mind) but if not this should work.
> * Over the modal "usubscribe mutliple" i can leave courses even if im the last administrator. This was already an issue in the past ([Filter courses and groups on "right to leave" in unsubscribe multiple view #4649](https://github.com/ILIAS-eLearning/ILIAS/pull/4649)) i assume this pr missed the fix.
> * When i deactivate "my coruses and groups" the shown link in the dashboard settings send me to the main menue settings, but the title of the page is "dashboard"
> 
> Greetings, @iszmais

I resolved these issues

fhelfer committed:
copyright js

iszmais reviewed (changes requested):
@fhelfer Thanks for your improvements!
Just one small request from my site to that:
- i would ask to add the initial settings on the persitence layer instead of as fallabck inside the ui. Im not quite sure where those are saved, but i assume withing the `settings` table (or another dashboard specific setting table)?
I assume this could be done with a db update step within the dashboard. Also i was still missing some settings in the current state (Learningseq and Studyprogramme are disabled by default and sortations are limited on list only. If this is no FR requirement i guess the default should be "all allowed").

Otherwise LGTM :+1:

iszmais commented on Services/Dashboard/Block/classes/class.ilDashboardBlockGUI.php:
```diff
@@ -567,6 +567,12 @@ public function manage(ReplaceSignal $replace_signal = null): string
             foreach ($item_group as $item) {
                 if ($this->rbacsystem->checkAccess('leave', $item->getRefId())) {
+                    if ($item->getType() === 'crs') {
+                        $mebersObj = ilParticipants::getInstance($item->getRefId());
```
@fhelfer And you might want to rename this variable :wink:

mjansenDatabay replied:
And: Variables in new code should also be `snake_case`.

fhelfer committed:
dbupdate dashboard initial settings

fhelfer committed:
dbinsert instead of isert ignore

iszmais reviewed (changes requested):
So far so good.

But still missing the "generall settings" :smile:

fhelfer committed:
copyright dbupdate dashboard

fhelfer committed:
update steps dashboard def pres/sort

iszmais commented:
Hello @kergomard 

We fixed your concerns within the PR.
Is there anything else additional to that?
If not i would appreciate your approval!

Greetings,
@iszmais

fhelfer committed:
roadmap updates

iszmais commented:
Hello @klees 

we updated the **Roadmap** to commit to the TBs concept in future improvements (which we are already working on).
Further im confident that the unification done by this PR will help to implement said desired changes.

Feel free to leave your feedback on that or your approval.

Greetings,
@iszmais

iszmais commented:
Hello,

 @fhelfer 
Due to time passed we got some merge conflicts here again. Could you resolve those?

@alex40724 
@smeyer-ilias 
@klees 
@kergomard 
To keep that our last time related issue i would **highly** appreciate you approvals for this PR (or alternativly any contribution in any way).

Greetings,
@iszmais

fhelfer committed:
Merge branch 'trunk' into refactor/global-dashboard-settings

fhelfer committed:
merge problems

alex40724 commented:
If possible, try to avoid to add another dependency to the yahoo library. We might get rid of it before getting rid of jquery. A fetch API call might be the best option currently.

kergomard reviewed (approved):
HI @fhelfer and @iszmais

Sorry, I somehow read  "Dashboard" and feel that it doesn't concern me.

The changes in `Object` are fine with me, thank you!
@kergomard

klees reviewed (approved):
I voiced my concerns and I'm not really happy still. But I wholeheartedly feel that we should [try to distribute responsibility and accountability](https://docu.ilias.de/goto_docu_copa_9682.html), so I guess this is your decision and you should move on =)

fhelfer committed:
Merge branch 'trunk' into refactor/global-dashboard-settings

smeyer-ilias reviewed (approved):
Thanks, the changes are fine for me.

iszmais commented:
@fhelfer
Thank you for your input and patience in this case!
On behalf of the ILIAS Society I apologize for the overexertion of such.

sincerely,
@iszmais
